nebuchadnezzar/nebu/h5p_injection.py

```diff
@@ -0,0 +1,314 @@
+        answers=answers,
+        is_answer_order_important=not random_answers
+    )
+
+
+def _true_false_question_factory(id: int, entry: dict[str, Any]):
+    behavior = entry.get("behaviour", {})
+    answers = []
+    parsed_correctness = try_parse_bool(entry["correct"])
+    for index, option in enumerate((True, False)):
+        is_correct = parsed_correctness == option
+        feedback_key = (
+            "feedbackOnCorrect"
+            if is_correct
+            else "feedbackOnWrong"
+        )
+        answers.append(
+            _answer_factory(
+                id=index + 1,
+                content_html=str(option),
+                correctness=is_correct,
+                feedback_html=behavior.get(feedback_key, ""),
+            )
+        )
+    return _question_factory(
+        id=id,
+        stem_html=entry["question"],
+        answers=answers,
+        is_answer_order_important=True,
+    )
+
+
+class SupportedLibrary(NamedTuple):
+    get_formats: Callable[[dict[str, Any]], List[str]]
+    make_question: Callable[[int, dict[str, Any]], dict[str, Any]]
+
+
+SUPPORTED_LIBRARIES = {
+    "H5P.MultiChoice": SupportedLibrary(
+        get_formats=lambda m: (
+            ["free-response", "multiple-choice"]
+            if m.get("is_free_response_supported", False)
+            else ["multiple-choice"]
+        ),
+        make_question=_multichoice_question_factory,
+    ),
+    "H5P.TrueFalse": SupportedLibrary(
+        get_formats=lambda _: ["true-false"],
+        make_question=_true_false_question_factory,
+    ),
+}
+
+
+def _add_question(
+    id: int,
+    library: str,
+    entry: dict[str, Any],
+    metadata: dict[str, Any],
+    questions: List[dict[str, Any]] = [],
+):
+    question = {}
+    if library in SUPPORTED_LIBRARIES:
+        lib = SUPPORTED_LIBRARIES[library]
+        question["formats"] = lib.get_formats(metadata)
+        question.update(**lib.make_question(id, entry))
+        question["collaborator_solutions"] = _make_collaborator_solutions(
+            entry
+        )
+        if entry.get("isSolutionPublic", None) is not True:
+            question["answers"] = [
+                {
+                    k: v
+                    for k, v in answer.items()
+                    if k not in ("correctness", "feedback_html")
+                }
+                for answer in question["answers"]
+            ]
+            question["collaborator_solutions"] = []
+        questions.append(question)
+    elif library == "H5P.QuestionSet":
+        for i, q in enumerate(entry["questions"]):
+            sub_library = q["library"].split(" ")[0]
+            sub_entry = q["params"]
+            assert sub_library != "H5P.QuestionSet", \
+                "Question sets cannot contain question sets"
+            _add_question(i + 1, sub_library, sub_entry, metadata, questions)
+    else:
+        raise UnsupportedLibraryError(library)
+
+
+def questions_from_h5p(nickname: str, h5p_in: dict[str, Any]):
```

Comment:
When is this used? I couldn't find it being called except in tests and trying to understand when we would construct an H5P Question from scratch in enki

Reply:
This is the main entry point for constructing questions from h5p content. It is used in the `interactive_callback_factory` inside `formatters.py` to add questions to the exercise dict.